pkg/checks/traceroute/traceroute.go

```diff
@@ -109,105 +149,159 @@ func readIcmpMessage(ctx context.Context, icmpListener *icmp.PacketConn, timeout
-	// since the channels size would be zero, blocking all threads from sending
+	// since the channel's size would be zero, blocking all threads from sending
 	queueSize := cfg.MaxHops * (1 + cfg.Rc.Count)
 	results := make(chan Hop, queueSize)
 	var wg sync.WaitGroup
 
 	for ttl := 1; ttl <= cfg.MaxHops; ttl++ {
+		c, hopSpan := tracer.Start(ctx, addr.String(), trace.WithAttributes(
+			attribute.Int("ttl", ttl),
+		))
 		wg.Add(1)
 		go func(ttl int) {
 			defer wg.Done()
+			defer hopSpan.End()
+
 			l := log.With("ttl", ttl)
-			logctx := logger.IntoContext(ctx, l)
-			err := helper.Retry(func(ctx context.Context) error {
-				hop, err := traceroute(ctx, addr, ttl, cfg.Timeout)
+			logctx := logger.IntoContext(c, l)
+
+			retry := 0
+			err = helper.Retry(func(ctx context.Context) error {
+				defer func() {
+					retry++
+				}()
+				hopSpan.AddEvent("Attempting to hop", trace.WithAttributes(
+					attribute.Int("ttl", ttl),
+					attribute.Int("retry", retry),
+				))
+
+				hop, hErr := hop(ctx, addr, ttl, cfg.Timeout)
 				if hop != nil {
 					results <- *hop
 				}
-				if err != nil {
-					l.Error("traceroute failed", "err", err.Error())
-					return err
+				if hErr != nil {
+					l.ErrorContext(ctx, "Failed to hop", "err", hErr)
+					hopSpan.SetStatus(codes.Error, hErr.Error())
+					hopSpan.RecordError(hErr)
+					return hErr
 				}
+
 				if !hop.Reached {
-					l.Debug("failed to reach target, retrying")
+					hopSpan.SetName(hop.Addr.String())
+					l.DebugContext(ctx, "Failed to reach target, retrying")
 					return errors.New("failed to reach target")
 				}
 				return nil
 			}, cfg.Rc)(logctx)
 			if err != nil {
-				l.Debug("traceroute could not reach target")
+				l.DebugContext(ctx, "Traceroute could not reach target")
+				if !errors.Is(err, syscall.EHOSTUNREACH) {
+					hopSpan.SetStatus(codes.Error, err.Error())
+					hopSpan.RecordError(err)
+				}
+				return
 			}
+			hopSpan.SetStatus(codes.Ok, "Hop succeeded")
 		}(ttl)
 	}
 
 	wg.Wait()
 	close(results)
 
+	// Collect and log hops
 	for r := range results {
 		hops[r.Ttl] = append(hops[r.Ttl], r)
 	}
+	logHops(ctx, hops)
 
-	printHops(ctx, hops)
-
+	sp.AddEvent("TraceRoute completed", trace.WithAttributes(
+		attribute.Int("hops_count", len(hops)),
+	))
 	return hops, nil
 }
 
-func ipFromAddr(remoteAddr net.Addr) net.IP {
-	switch addr := remoteAddr.(type) {
-	case *net.UDPAddr:
-		return addr.IP
-	case *net.TCPAddr:
-		return addr.IP
-	case *net.IPAddr:
-		return addr.IP
-	}
-	return nil
-}
-
-// traceroute performs a traceroute to the given address with the specified TTL and timeout.
+// hop performs a hop to the given address with the specified TTL and timeout.
 // It returns a Hop struct containing the latency, TTL, address, and other details of the hop.
-func traceroute(ctx context.Context, addr net.Addr, ttl int, timeout time.Duration) (*Hop, error) {
+func hop(ctx context.Context, addr net.Addr, ttl int, timeout time.Duration) (*Hop, error) {
```

Comment:
Shouldn't we rename this to `doHop` or something similar to avoid the shadowing with the `hop` variable throughout the whole file? Or just use `res` or `h` for the hop in functions?

A conflict seems kind of impossible, but why even take the risk?